wpcom-legacy-redirector.php

```diff
@@ -104,13 +105,58 @@ static function insert_legacy_redirect( $from_url, $redirect_to ) {
 			return new WP_Error( 'invalid-redirect-url', 'Invalid redirect_to param; should be a post_id or a URL' );
 		}
 
-		wp_insert_post( $args );
+		switch ( $mode ) {
+			case 'update':
+				$args[ 'ID' ] = self::get_redirect_post_id( $from_url, false );
+				if ( 0 === wp_update_post( $args ) ) {
+					return new WP_Error( 'failed-redirect-update', 'Redirect update failed for unknown reason' );
+				}
+				break;
+			case 'add':
+			default:
+				wp_insert_post( $args );
+				break;
+		}
 
 		wp_cache_delete( $from_url_hash, self::CACHE_GROUP );
 
 		return true;
 	}
 
+	/**
+ 	 * @param string $from_url URL or path for the redirect we'd like to delete.
+ 	 * @return bool|WP_Error WP_Error if something is amiss with the delete; true otherwise.
+ 	 */
+
+	static function delete_legacy_redirect( $from_url ) {
+		if ( ! ( defined( 'WP_CLI' ) && WP_CLI ) && ! is_admin() ) {
+			// never run on the front end
+			return false;
+		}
+
+		$from_url = self::normalise_url( $from_url );
+		
+		if ( is_wp_error( $from_url ) ) {
+			return false;
+		}
+
+		$post_id = self::get_redirect_post_id( $from_url, false);
+
+		//sanity check that we are deleting only the correct post_type
+		if ( get_post_type( $post_id ) !== self::POST_TYPE ) {
+			return new WP_Error( 'failed-redirect-delete', 'Redirect delete attempted to delete a post of incorrect type.' );
+		}
+
+		if ( 0 !== $post_id ) {
+			wp_delete_post( $post_id, true );
+		}else{
+			return false;
+		}
+
+		wp_cache_delete( self::get_url_hash( $from_url ), self::CACHE_GROUP );
+		return true;
```

Comment:
The function does return true anytime it attempted to delete the post and flush cache. But it does not mean the post has been deleted. `wp_delete_post` returns `false` on failure. Perhaps we should check whether it succeed or not instead of blindly trusting that it has succeeded?